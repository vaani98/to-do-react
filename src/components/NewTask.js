import React from 'react';
import './NewTask.css'
const NewTask = props =>{
    return(
        // Can use Fragments here, it's helpful when we're operating on a large app, saves an html DOM node
        <div> 
        <label htmlFor="new-task">Add Item</label><br/>
        <form onSubmit={(e)=>props.addItem(e)}>
            <input id ="new_item"
            type="text"
            className="input"
            placeholder="Add a new task"
            style={{border: '1px solid #ddd'}}
            />
            <img 
            src="https://img.icons8.com/color/96/000000/add.png" 
            alt = "Add Task" className = "icon-btns" 
            id = "add-btn" 
            onClick = {(e)=>props.addItem(e)}>
            </img>
        </form>
        </div>
    )
}
export default NewTask;
