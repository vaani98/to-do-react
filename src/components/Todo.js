import React, {useState, useEffect} from 'react';
import './Todo.css';
import Item from './Item'
import NewTask from './NewTask'
import ErrorModal from './ErrorModal'

const persistData = tasks =>{
    localStorage.setItem('todo', JSON.stringify(tasks));
}    
const readStorage = () =>{
    const storage = JSON.parse(localStorage.getItem('todo'));
    if(storage){
        for(var i=0; i<storage.length; i++){
            storage[i].date = new Date(storage[i].date)
        }
        return storage;
    }
    else 
        return []
}

const Todo = () =>{

    // defined these in the beginning of the function. 
    const [tasks, setTasks] = useState([]);
    const [errorText, setErrorText] = useState(null);
    //removed tasksRemaining variable
    //moved persistData and readStorage outside the component 
    
    useEffect(() => {
        const tasksRead = readStorage();
        setTasks(tasksRead);
    }, []);

    useEffect(() => {
        persistData(tasks);
        
    }, [tasks]);
    
    const clearInput = () =>{
        document.getElementById("new_item").value = '';
    }

    const addItem = (e) => {
        e.preventDefault();
        const itemText = document.getElementById("new_item").value;
        var titles = tasks.map(el => el.title);    
        var duplicate = titles.includes(itemText);  
        //added logic to avoid adding duplicates, empty tasks, and tasks that are longer than 20 characters, setting error text accordingly  
        if(!duplicate && itemText.length<=20 && itemText.length>0){
            let newTaskItem = {
                title:  itemText,
                completed: false, 
                date: new Date()
            };
            // populate your original data and then added the new data
            let newTasks = [...tasks];
            //add to beginning of array 
            newTasks.unshift(newTaskItem);
            setTasks(newTasks);    
        }
        else if(itemText.length>20){
            setErrorText("task too long to add! (max 20 characters)");
        }
        else if(duplicate){
            setErrorText("can't add duplicate task!");
        }
        else{
            setErrorText("can't add empty task!");
        }
        clearInput();
    }

    const clearError = () =>{
        setErrorText(null);
    }

    const toggleComplete = index => {
        const newTasks = [...tasks];
        newTasks[index].completed = !newTasks[index].completed
        setTasks(newTasks);
    }

    const deleteItem = (index) =>{
        const newTasks = [...tasks];
        newTasks.splice(index, 1);
        setTasks(newTasks);
    }

    const sortDate = () =>{
        console.log(tasks);
        let tasksCopy = [...tasks];
        let newTasks = tasksCopy.sort((a, b) => b.date - a.date);
        setTasks(newTasks);
    }

    const sortAlpha = () =>{
        let tasksCopy = [...tasks];
        let newTasks = tasksCopy.sort((a, b) => a.title.localeCompare(b.title))
        setTasks(newTasks);
    }

    return(
        <div className="container">
            {errorText && <ErrorModal onClose = {clearError}>{errorText}</ErrorModal>}
            <NewTask addItem = {addItem}></NewTask>
            <div className="tasks">
                <h3>Todo</h3>
                <button 
                className = "sort" 
                id = "sort_date" 
                onClick = {sortDate}>
                    Sort by Date
                </button>
                <button 
                className = "sort" 
                id = "sort_alpha" 
                onClick = {sortAlpha}>
                    Sort Alphabetically
                </button>
                <ul className = "todo">
                {tasks.map((task, index) => (
                    <Item
                        title={task.title}
                        completed = {task.completed}
                        index={index}
                        key={index}
                        deleteTask = {deleteItem}
                        toggle = {toggleComplete}
                    />
                ))}
                </ul>
            </div>
        </div>
    )
} 

export default Todo;
