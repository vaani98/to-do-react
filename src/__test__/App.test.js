import React from 'react';
import {render, cleanup, fireEvent, getByAltText, getByText, queryByText, toHaveStyle, getByPlaceholderText} from '@testing-library/react'

import '@testing-library/jest-dom/extend-expect'
import Todo from '../components/Todo';
import NewTask from '../components/NewTask';

describe('my test suite', ()=>{

    afterEach(cleanup);

    it('renders', () => {
        const {asFragment} = render(<Todo/>);
        expect(asFragment()).toMatchSnapshot();
    });
    
    it('enters input', () =>{
        const renderDOM = render(<NewTask/>);
        let newText = 'Get Milk';
        const inputElement = renderDOM.getByPlaceholderText('Add a new task');
        inputElement.value = newText;
        fireEvent.change(inputElement);
        expect(inputElement.value).toBe('Get Milk');
    });
    
    it('calls add item function', () => {
        const addItem = jest.fn();
        const renderDOM = render(<NewTask addItem = {addItem} />)
        let newText = 'Get Milk'
        fireEvent.change(renderDOM.getByPlaceholderText('Add a new task'), {target: {value: newText}})
        renderDOM.getByAltText('Add Task').click()  
        expect(addItem).toHaveBeenCalledTimes(1);
    });
    
    
    it('adds item to DOM ul', () => {
        const reactDOM = render(<Todo/>);
        const taskList = document.querySelector('.todo');
        const numberOfItems = taskList.children.length;
        let newText = 'Get Milk';
        fireEvent.change(reactDOM.getByPlaceholderText('Add a new task'), {target: {value: newText}});
        reactDOM.getByAltText('Add Task').click();
        expect(taskList.children.length).toBe(numberOfItems + 1);
    });
    
    it("doesn't add empty task", () => {
        const reactDOM = render(<Todo/>);
        let newText = '';
        const taskList = document.querySelector('.todo');
        const numberOfItems = taskList.children.length;
        fireEvent.change(reactDOM.getByPlaceholderText('Add a new task'), {target: {value: newText}});
        reactDOM.getByAltText('Add Task').click();
        expect(taskList.children.length).toBe(numberOfItems);
    })
    
    it('adds right text', () => {
        const reactDOM = render(<Todo/>);
        let newText = 'Buy Bread';
        fireEvent.change(reactDOM.getByPlaceholderText('Add a new task'), {target: {value: newText}});
        reactDOM.getByAltText('Add Task').click();
        const taskList = document.querySelector('.todo');
        expect(taskList.children[0].textContent).toBe(newText);
    })
        
    it("doesn't add duplicate tasks", () => {
        const reactDOM = render(<Todo/>);
        let newText = 'Brush Teeth';
        const taskList = document.querySelector('.todo');
        const numberOfItems = taskList.children.length;
        fireEvent.change(reactDOM.getByPlaceholderText('Add a new task'), {target: {value: newText}});
        reactDOM.getByAltText('Add Task').click();

        fireEvent.change(reactDOM.getByPlaceholderText('Add a new task'), {target: {value: newText}});
        reactDOM.getByAltText('Add Task').click();
        expect(taskList.children.length).toBe(numberOfItems + 1);
    })

    it("deletes items", () =>{
        const reactDOM = render(<Todo/>);
        let newText = 'Go to the doctor';
        fireEvent.change(reactDOM.getByPlaceholderText('Add a new task'), {target: {value: newText}});
        reactDOM.getByAltText('Add Task').click();
        
        const taskList = document.querySelector('.todo');
        const newItem = taskList.children[0];
        var deleteBtn = getByAltText(newItem, 'delete');
        deleteBtn.click();

        expect(queryByText(taskList, newText)).toBeNull();
    })

    it("toggles strikethrough property", () =>{
        const reactDOM = render(<Todo/>);
        let newText = 'Take a nap';
        fireEvent.change(reactDOM.getByPlaceholderText('Add a new task'), {target: {value: newText}});
        reactDOM.getByAltText('Add Task').click();        
        const taskList = document.querySelector('.todo');
        const newItem = taskList.children[0];
        var checkbox = newItem.querySelector("input[type='checkbox']");
        checkbox.click();
        expect(queryByText(taskList, newText)).toHaveStyle(`text-decoration: line-through`);
        checkbox.click();
        expect(queryByText(taskList, newText)).toHaveStyle(`text-decoration: `);
    })

});
