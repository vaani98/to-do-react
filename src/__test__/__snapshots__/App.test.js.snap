// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`my test suite renders 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <label
      for="new-task"
    >
      Add Item
    </label>
    <br />
    <form>
      <input
        class="input"
        id="new_item"
        placeholder="Add a new task"
        style="border: 1px solid #ddd;"
        type="text"
      />
      <img
        alt="Add Task"
        class="icon-btns"
        id="add-btn"
        src="https://img.icons8.com/color/96/000000/add.png"
      />
    </form>
    <div
      class="tasks"
    >
      <h3>
        Todo
      </h3>
      <button
        class="sort"
        id="sort_date"
      >
        Sort by Date
      </button>
      <button
        class="sort"
        id="sort_alpha"
      >
        Sort Alphabetically
      </button>
      <ul
        class="todo"
      />
    </div>
  </div>
</DocumentFragment>
`;
